kit/scrolls/src/Lists/UICollapsableSectionList/UICollapsableSectionList.tsx

```diff
@@ -0,0 +1,323 @@
+} & SectionListProps<ItemT, SectionT>) {
+    const [foldedSections, setFoldedSections] = React.useState<Record<string, boolean>>({});
+
+    const processedSections = React.useMemo(() => {
+        return sections.map(section => {
+            const { key } = section;
+            if (!key) {
+                return section;
+            }
+            if (foldedSections[`${key}:header`]) {
+                return {
+                    ...section,
+                    data: emptyArray as ItemT[],
+                };
+            }
+            return section;
+        });
+    }, [sections, foldedSections]);
+
+    const renderCollapsableSectionHeader = React.useCallback(
+        (info: { section: SectionListData<ItemT, SectionT> }) => {
+            if (info.section.key == null) {
+                // TODO: can we do sth with it or better to throw an error?
+                return null;
+            }
+            const sectionKey = `${info.section.key}:header`;
+            return (
+                <TouchableOpacity
+                    onPress={async () => {
+                        now = Date.now();
+
+                        prepareAnimation(
+                            sectionKey,
+                            foldedSections,
+                            screenshotRef,
+                            listRef,
+                            sectionsMapping,
+                            sectionToAnimateKey,
+                        );
+
+                        setFoldedSections({
+                            ...foldedSections,
+                            [sectionKey]: !foldedSections[sectionKey],
+                        });
+                    }}
+                >
+                    {renderSectionHeader?.(info)}
+                </TouchableOpacity>
+            );
+        },
+        [
+            renderSectionHeader,
+            foldedSections,
+            listRef,
+            screenshotRef,
+            sectionsMapping,
+            sectionToAnimateKey,
+        ],
+    );
+
+    return (
+        <AccordionOverlayView ref={screenshotRef} style={rest.style}>
+            <VirtualizedSectionList
+                ref={listRef}
+                {...rest}
+                sections={processedSections}
+                // extraData={processedSections.reduce(
+                //     (acc, { data }) => acc + data.length,
+                //     0,
+                // )}
+                renderSectionHeader={renderCollapsableSectionHeader}
+            />
+        </AccordionOverlayView>
+    );
+}
+
+export function UICollapsableSectionList<ItemT, SectionT = DefaultSectionT>(
+    props: SectionListProps<ItemT, SectionT>,
+) {
+    const { sections, contentContainerStyle } = props;
+
+    const sectionsMapping = React.useRef<Record<string, string>>({});
+    const prevSections = React.useRef<typeof props['sections']>().current;
+
+    const sectionToAnimateKey = React.useRef<string | undefined>();
+
+    if (prevSections !== sections) {
```

Comment:
What do you of memoizing these calculations?

Reply:
Not sure that I understood your question.
It's actually kind of memoization. Regarding the passed sections - we can't do anything, it's a user responsibility to cache it.

Reply:
Hm… actually you are right, it’s a rare action indeed due to pointers to section objects comparison